Social Login: Allow logging into an existing WP.com account with a Google account

pento opened the pull request:
This pull request addresses #15542.

If the API attempts to create a social user with the same email as an existing WP.com account, it will return a `user_exists` error.

We can catch that error and prompt the user to login to their WP.com account, then link their social account.

#### Testing instructions

1. Run `git checkout feature/15542-social-login-with-existing-account` and start your server, or open a [live branch](https://calypso.live/?branch=feature/15542-social-login-with-existing-account)
2. If you don't have one already, create a normal WordPress.com account with the same email address as a Google account through the normal register process. Don't user Social Login for this.
3. ~~Apply D6171 to your sandbox, and sandbox wordpress.com.~~ (D6171 has been deployed.)
4. Open the [`Log In` page](http://calypso.localhost:3000/log-in)
5. Use Social Login with the matching Google account.

#### Reviews

- [x] Code
- [x] Product
- [x] Tests

pento committed:
Login: Initial work on allowing social login to an existing account

pento committed:
Display the login form correctly after an existing WP.com user is found.

pento committed:
Remove bonus whitespace.

fabianapsimoes commented:
I don't see the second screen in #15542. After confirming my password, I'm logged in and redirected to the Reader directly. If I try to log in again, I see the prompt to connect my account again (_"We found a WordPress.com account with the email address..."_). I've tested with the accounts `fabianatestingsignup2` and `fabianareader`.

Using an account created via `/start/social` works fine. I don't see that prompt and am logged in directly.

pento commented:
The second screen only exists in some hacky code on my computer right now, hence the `In Progress` label. 🙂

Ran into some problems dealing with `state` today, hopefully I'll have a usable commit tomorrow.

fabianapsimoes commented:
> The second screen only exists in some hacky code on my computer right now, hence the In Progress label. :slightly_smiling_face:

Oh, it was on Needs Review in our board tomorrow so I figured you just had forgotten to change the labels here. Updating it back to `In Progress`.

pento committed:
A currently non-functional attempt at adding the Social Connect prompt screen

pento commented:
Let's Play: Redux code written by someone who doesn't know Redux!

152bdd4 is the current state of what I've been hacking away at for the past few days. There's probably a lot of cruft left over from these hacks, so I expect a lot of it to look like gibberish. 🙃

To give a brief overview:

I initially started writing it as a fairly simple thing - after the API returned that the user already existed, it would change the `LoginForm` to the new style. After successfully logging in, it would switch to the `SocialConnectPrompt`. I ran into a problem here where it seems that the user data isn't quite refreshed properly after login, hence the hard refresh of `Login.rebootAfterLogin()`.

With that in mind, I tried making the `SocialConnectPrompt` the first step of the reboot - send the user to `/log-in/social-connect`, show the correct screen, tada! Unfortunately, that also had problems - it couldn't be server side rendered, so the display flashed between the login form, then the connect prompt. I also found out that `state` doesn't survive a refresh, so it didn't have any information about the social user.

And so, I decided to try a Redux-based solution, trying to copy how 2FA does it. Unfortunately, I don't really understand how it works, so it appears I've missed some critical part of how it all ties together. There's a lot of redundant flags and code in the `Login` block at the moment, the current problem is that, after logging into the WordPress.com account, `Login.render()` isn't run, to show the `SocialConnectPrompt`. I've tried various forms of updating the state, I even tried a `forceUpdate()`, but nothing seems to be working there.

And that brings us to now. 🙂

pento commented:
Rebased.

pento committed:
Remove renderError(), accidentally included after rebasing.

pento committed:
Remove a bunch of superfluous state management

pento commented:
So, it turns out that the problems were due to bugs in the server side code, not the client side. 🙃

I've started cleaning up all the mess I added while trying to debug the wrong problem, there's still a bunch to go, though.

scruffian commented:
We need to remember that sometimes an email address won't be the primary address on the account, so users won't be able to log in with it, even though they won't be able to use it.

pento committed:
Remove unnecessary forceUpdate()

pento committed:
Disable the Google button and login form while requesting a social login

pento committed:
Add styling to SocialConnectPrompt

pento committed:
Remove an unnecessary selector from SocialConnectPrompt

pento committed:
Capitalize the social service name in SocialConnectPrompt

pento committed:
Capitalize the service name in the header, too.

pento committed:
LoginForm: Hide the Remember Me button when linking a social account.

pento committed:
Add handlers for connecting the social account from the SocialConnectPrompt

scruffian commented:
It might be good to squash some commits before this is merged.

scruffian commented on client/blocks/login/error-notice.jsx:
```diff
@@ -57,7 +57,7 @@ class ErrorNotice extends Component {
 		const error = this.getError();
 
-		if ( ! error || ( error.field && error.field !== 'global' ) || ! error.message ) {
+		if ( ! error || ( error.field && error.field !== 'global' ) || ! error.message || error.code === 'existing_wpcom_user' ) {
```
This is quite a long condition. Maybe it would be nice to move it to a function?

pento replied:
I've split the error.code check into a separate condition, as it deserves a comment to explain what's happening.

scruffian commented on client/blocks/login/social-connect-prompt.jsx:
```diff
@@ -0,0 +1,112 @@
+			<Card className="login__social-connect-prompt">
+				<div className="login__social-connect-prompt-logos">
+					{ /* eslint-disable max-len */ }
+					<svg className="login__social-connect-prompt-logo" viewBox="-2 -2 24 24" xmlns="http://www.w3.org/2000/svg">
```
I think we should move this to an image.

pento replied:
See Automattic/social-logos#46.

scruffian commented on client/login/wp-login/index.jsx:
```diff
@@ -70,14 +80,19 @@ export class Login extends React.Component {
 			<div>
 				<Main className="wp-login__main">
-					{ this.renderLocaleSuggestions() }
+					{ ! socialConnect &&
```
Should we also do this on 2FA pages?

pento replied:
This check is redundant, it's handled in `renderLocaleSuggestions()` (along with 2FA pages).

scruffian commented on client/state/login/actions.js:
```diff
@@ -278,7 +283,40 @@ export const createSocialUser = ( service, token, flowName ) => dispatch => {
+		dispatch( {
+			type: SOCIAL_CONNECT_ACCOUNT_REQUEST_SUCCESS,
+			redirectTo: wpcomResponse.redirect_to,
+		} );
```
Should we return a `Promise` here too?

pento replied:
I don't think so, none of the success routines in the other login functions return a `Promise`.

scruffian commented:
I tested this but I get this error:
`An active access token must be used to add a social login account to the current user.`

pento commented:
I'll tackle the other bits on Monday, but I wanted to note for this:

> An active access token must be used to add a social login account to the current user.

I briefly mentioned the problem I've run into on the Trello card, this is the symptom - the new `/me/social-login/connect` endpoint requires the user be logged in, but Calypso doesn't send the user token for API requests until it's been reloaded. 

Unfortunately, we can't reload until after the wpcom account and social account are linked, because the social account token is stored in the state, which is lost on reload. I'm not wild about storing it anywhere more permanent, as it will probably cause a mess of security issues.

So, I'd really like it if there was a magic way to trick Calypso into sending the user token with API requests, before it's reloaded.

scruffian commented:
> So, I'd really like it if there was a magic way to trick Calypso into sending the user token with API requests, before it's reloaded.

I'm looking into it....

pento committed:
Experiments with reloading the proxy iframe.

Note: This branch is broken without an experimental version of
`wpcom-proxy-request`, that I haven’t committed yet.

scruffian committed:
remove files

scruffian committed:
remove reloadproxy

scruffian committed:
remove these changes so that i can use the branch

pento committed:
Switch back to the wpcom-proxy-request reload branch

pento committed:
Split the ErrorNotice display condition, to explain why existing_wpcom_user is checked.

pento committed:
Remove a redundant socialConnect page check.

pento commented:
Hooray, it's finally working! A few notes for when you're testing:

- Please ensure you've done a hard reset on your local copy - the branch has been rebased a couple of times, so it may be in a funky state.
- You _must_ run `npm upgrade` manually, to ensure the correct version of `wpcom-proxy-request` is installed. See [here](https://github.com/Automattic/wpcom-proxy-request/compare/feature/reload) for what's new.
- D6171 has been recently updated, please ensure you have the latest copy on your sandbox.

pento committed:
Don't reset the social login state when the SOCIAL_CONNECT_ACCOUNT_REQUEST_SUCCESS action fires.

This was causing some UI jittering when the state updated, but the page
hadn’t finished reloading.

yurynix commented on client/state/login/actions.js:
```diff
@@ -278,7 +283,40 @@ export const createSocialUser = ( service, token, flowName ) => dispatch => {
+ * @param  {String}    redirectTo Url to redirect the user to upon successful login
+ * @return {Function}             Action thunk to trigger the login process.
+ */
+export const connectSocialUser = ( service, token, redirectTo ) => dispatch => {
```
I think new endpoints should live under the new data layer, like so:
https://github.com/Automattic/wp-calypso/blob/master/client/state/data-layer/wpcom/me/notification/settings/index.js

But maybe we should do that in a follow up PR.

yurynix commented on client/login/wp-login/index.jsx:
```diff
@@ -48,13 +53,18 @@ export class Login extends React.Component {
 
+		if ( socialConnect ) {
+			url += `/${ socialConnect }`;
+			title += ' > Social Connect';
```
Missing localization ?

yurynix replied:
ohh, it's just for the pageview, nevermind

yurynix commented on client/lib/wpcom-undocumented/lib/me.js:
```diff
@@ -361,6 +366,38 @@ UndocumentedMe.prototype.deletePurchase = function( purchaseId, fn ) {
+		body: body,
+	};
+
+	require( 'wpcom-proxy-request' ).reloadProxy();
```
Maybe it's worth documenting why proxy reload is required in that case

pento replied:
👍🏻

yurynix commented on client/blocks/login/index.jsx:
```diff
@@ -57,22 +63,28 @@ class Login extends Component {
-		if ( ! this.props.twoFactorEnabled ) {
-			this.rebootAfterLogin();
-		} else {
+		if ( this.props.twoFactorEnabled ) {
```
I'm maybe missing something here, but once the user gone to twoFactor screens, how he's getting back to the linking social screen?

pento replied:
Good catch, fixed in 4d73246.

pento committed:
Add a comment explaining why the proxy reload happens.

pento committed:
Ensure that 2FA users are redirected to the Social Connect screen.

pento committed:
Disable the Google button until the Google API client has initialised.

We have unacceptable behaviour if it’s clicked too early:

- If it’s clicked before the login form is finished loading, the form
tries to POST to /log-in.
- If it’s clicked before the client is initialised, we get a “popup
blocked” error in the browser, and the user needs to click it again.

pento committed:
Move login form disabled state to a reducer/selector, so it can be reused.

pento committed:
Remove an unnecessary div

pento committed:
Login state: Add tests for modified/added reducers and selectors.

yurynix commented:
I think there's a missing step for `wpcom-proxy-request`, it's not enough to have the branch in the `package.json`, we also need the build artifact...

```
git clone git@github.com:Automattic/wpcom-proxy-request.git
cd wpcom-proxy-request
npm install
cp build/index.js* ../wp-calypso/node_modules/wpcom-proxy-request/build/ 
```

yurynix commented:
even with this, I get no errors in console but:
![screen shot 2017-07-17 at 16 00 44](https://user-images.githubusercontent.com/326402/28269035-2dfdee38-6b09-11e7-9131-ad8687ff49ee.png)

and the login looks like that if I navigate directly:
![screen shot 2017-07-17 at 16 01 53](https://user-images.githubusercontent.com/326402/28269055-46c3ed46-6b09-11e7-8fda-e6dad295a79a.png)

yurynix commented:
Screenshots:
<img width="495" alt="screen shot 2017-07-17 at 19 02 08" src="https://user-images.githubusercontent.com/326402/28277407-8dddf9c4-6b22-11e7-890d-81ea5b3c005d.png">

<img width="423" alt="screen shot 2017-07-17 at 19 02 18" src="https://user-images.githubusercontent.com/326402/28277417-94e099de-6b22-11e7-8473-f8b5fcfd6e26.png">

<img width="467" alt="screen shot 2017-07-17 at 19 02 33" src="https://user-images.githubusercontent.com/326402/28277429-9bb3ddde-6b22-11e7-89ef-fba8552cb5f4.png">

yurynix commented:
Works as described! 👍  (server side needs a little change)

Tug committed:
Use wpcom-proxy-request v4.0.1 and update npm-shrinkwrap.json

Tug committed:
Use wpcom-proxy-request v4.0.2 and update npm-shrinkwrap.json

scruffian commented:
I tested and this works as expected. We still need to account for 2FA users and make sure they enter their secondary authentication, but we can do that in another PR.

danhauk committed:
Update the illustration for social connect prompt

danhauk commented:
I pushed an update to the illustration on the connect social prompt. The giant arrow was kind of awkward so I used some dots that are a little more subtle and in line with the new illustration style.

![image](https://user-images.githubusercontent.com/448298/28430945-6dd09062-6d50-11e7-9692-7bf5dd9f4d0b.png)

I also have some feedback on the design of the first screen where we ask the user to log in to connect.

1 - I'd like to draw attention to the copy above the login form. When I was testing this PR I nearly missed it since the only thing that has really changed is we added text above the form. All the form fields are still there as well as the Google button.

I imagine if a user isn't expecting it or looking for it like I was they would definitely miss it. It's jarring and confusing to think you're logging in with your Google account and then land on this login form. As a user, if I missed the text I would think it's a bug.

An option for clarifying this message is to use a notice above the form instead of text inside the card and update the text on the button to reflect the intent of this form.

2 - The Google button should probably be downplayed here. I'm making an assumption, but I don't imagine too many users will be logging in with the wrong Google account. It seems like an edge case that doesn't justify the prominence of the button, competing with the primary button. It also rectifies the issue of the form looking exactly the same as the previous step.

Proposed changes:

![image](https://user-images.githubusercontent.com/448298/28431160-45ce2600-6d51-11e7-959c-7b447d4e2729.png)

(Alternatively, for the Google button issue, we just remove the social section from the card completely and add a `← Back` link below the form.)

Happy to address this in another PR since this one is pretty big already.

Tug commented:
LGTM 👍 

Let's merge this and work on 2FA, social disconnect etc in other PRs.
